File: add-two-numbers/add-two-numbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    dummyHead := &ListNode{Val: 0}
    currentNode := dummyHead
    carry := 0

    for l1 != nil || l2 != nil {
        var x, y int
        if l1 != nil {
            x = l1.Val
            l1 = l1.Next
        }
        if l2 != nil {
            y = l2.Val
            l2 = l2.Next
        }
        sum := carry + x + y
        carry = sum / 10
        currentNode.Next = &ListNode{Val: sum % 10}
        currentNode = currentNode.Next
    }

    if carry > 0 {
        currentNode.Next = &ListNode{Val: carry}
    }

    return dummyHead.Next
}

File: add-two-numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        l1head = l1
        l2head = l2

        l1number = ""
        l2number = ""

        while l1head != None:
            # print(l1head.val)
            l1number += f"{l1head.val}"
            l1head = l1head.next

        while l2head != None:
            # print(l2head.val)
            l2number += f"{l2head.val}"
            l2head = l2head.next

        l1number = int(l1number[::-1])
        l2number = int(l2number[::-1])
        
        result = l1number + l2number

        result = str(result)[::-1]

        # print(result)

        rhead = ListNode(result[0])
        current_node = rhead
        for char in result[1:]:
            current_node.next = ListNode(int(char))
            current_node = current_node.next

        return rhead


